### PR TITLE
704 requistion graphql schema

### LINKED ISSUE #704: Requisition Graphql Schema
https://github.com/openmsupply/remote-server/issues/703

### PR DESCRIPTION
**Pull request by andreievg**
Closes #704 

`NOTE` none of the new schema is implemented

I've tried to consider both related issues: https://github.com/openmsupply/remote-server/issues/625, https://github.com/openmsupply/remote-server/issues/624, but did vary from it a bit.

I think overall issue for Order Cycle is good to check out before going into full review: https://github.com/openmsupply/remote-server/issues/703

### InvoiceNode

Added requisition and linkedShipment

Screen Shot 2022-01-26 at 12 44 45 AM

### ItemNode

Added stats
Screen Shot 2022-01-26 at 12 46 16 AM

### RequisitionNode

Screen Shot 2022-01-26 at 12 52 01 AM

Please see parent issue for statuses, https://github.com/openmsupply/remote-server/issues/703

### RequisitionLineNode
Screen Shot 2022-01-26 at 12 52 47 AM

`stats` and `requestStats` are snapshots of itemStats at the time of line creation

also please note `inbound/outboundShipmentLines`, instead of doing calculations on how much was picked and shipped, this is available now by exposing lines ?

**Question**: Should invoice be exposed on invoice_line ?  To figure out if quantity in line refers to picked/allocated/shipped/delivered etc.. ?

### Requisition queries

requisition and requisitionByNumber (with id or number), pretty straight forward,

for requisitions query:

Screen Shot 2022-01-26 at 12 58 49 AM

Screen Shot 2022-01-26 at 1 02 09 AM

### Mutations

Ended up going with Request and Response rather then Customer and Supplier, when me Richard and Craig had a go at figuring out a name for them, didn't really reach conclusion, so followed similar patter to invoice where mutation name was tied to type.

#### RequisitionLine

Only update available for Response (since we agreed to leave it out for phase 1). Item can only be set on insert and update only changes requested and supply (supply being `supply this invoice` in mSupply)

Calculated quantity is set when line is inserted, or when threshold_months_of_stock or max_months_of_stock is set on requisition.

Snapshots stats would be set when line is inserted

**Question** Do we need 'refresh' stats mutation ? To get current stats for item ?

#### Requisition

I think it's pretty straight forward, only update for Response.

Ended up doing something similar to Shipments for allowable status updates

#### Extras

Wasn't sure how to name the extra mutations

`useCalculatedQuantity` -> RequestRequisition only (set requested = calculated for all lines)

`addFromMasterList` -> RequestRequisition only

`supplyRequestedQuantity` -> ResponseRequisition only (set supply = requested for all lines)

`createRequistionShipment` -> ResponseRequisition only

I used input parameter names to give a bit more clarity about requisition types (i.e. responseRequistionId and requestRequisitionid),

for that last mutation, i am not sure how to calculate quantity (will check with mSupply when time comes to do it, for now it's as per schema comment)
Screen Shot 2022-01-26 at 1 22 45 AM

### Other errors

There is very limited amount of structured errors, only thought about possible conflicting errors (i.e. concurrent editing of requisition, or some setting was changed while requisition was being edited).

Other errors will be present as standard graphql error, for bad input (like when trying to do a Request mutation or Response requisition)

### CODE REVIEW

File: graphql/src/schema/queries/requisition.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    pub r#type: Option<EqualFilterInput<RequisitionNodeType>>,
+    pub status: Option<EqualFilterInput<RequisitionNodeStatus>>,
+    pub created_datetime: Option<DatetimeFilterInput>,
+    pub sent_datetime: Option<DatetimeFilterInput>,
+    pub finalised_datetime: Option<DatetimeFilterInput>,
+    pub other_party_name: Option<EqualFilterStringInput>,
+    pub other_party_id: Option<EqualFilterStringInput>,
+    pub color: Option<EqualFilterStringInput>,
+    pub their_reference: Option<EqualFilterStringInput>,
+    pub comment: Option<EqualFilterStringInput>,
```
Comment (clemens-msupply) [inline graphql/src/schema/queries/requisition.rs:44]: SimpleStringFilterInput for comment?

**Review by clemens-msupply — APPROVED** [review]
Looks good (with my limited knowledge on requisitions.

1) There might be a couple of lint warnings that can be fixed (unused imports)
2) With Mark already looking into store is we could probably start making them required?

File: graphql/src/schema/mutations/mod.rs [hunk 2]
```diff
@@ -366,6 +368,133 @@ impl Mutations {
...
+        ctx: &Context<'_>,
+        store_id: Option<String>,
+        input: request_requisition::UseCalculatedQuantityInput,
+    ) -> Result<request_requisition::UseCalculatedQuantityResponse> {
+        // TODO remove and make store_id parameter required
+        request_requisition::use_calculated_quantity(ctx, store_id, input)
+    }
+
+    /// Add i
+    async fn add_from_master_list(
```
[the remark this answers is not in the document]
Comment (andreievg) [reply graphql/src/schema/mutations/mod.rs:414]: thanks: 36e8bc39598d6bd7196d5ce6b185b06121320203

File: graphql/src/schema/mutations/requisition/request_requisition/insert.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    pub comment: Option<String>,
+    pub max_months_of_stock: f64,
+    pub threshold_months_of_stock: f64,
+}
+
+#[derive(Interface)]
+#[graphql(name = "InsertRequestRequisitionErrorInterface")]
+#[graphql(field(name = "description", type = "String"))]
+pub enum InsertErrorInterface {
+    OtherPartyNotASupplierStore(OtherPartyNotASupplierStore),
```
[the remark this answers is not in the document]
Comment (andreievg) [reply graphql/src/schema/mutations/requisition/request_requisition/insert.rs:21]: This is one use case where I thought it's possible for supplier -> store relation to change (it's central configuration, so could sync and change it while name is being searched)

File: graphql/src/schema/queries/requisition.rs (hunk 1 above)
Comment (andreievg) [reply graphql/src/schema/queries/requisition.rs:44]: keen to chat about this one in our meeting today

**Comment by andreievg** [comment]
> 1 - There might be a couple of lint warnings that can be fixed (unused imports)

I fixed them upstream, thanks

> 2 - With Mark already looking into store is we could probably start making them required?

Sure, i'll do it when i query/mutations are implemented

**Pull request merged**